### PR TITLE
chore(sql): fix backup SQL error when table is concurrently changed

### LINKED ISSUE #5201: BACKUP TABLE SQL fails with TableReferenceOutOfDateException
### To reproduce

Scenario:

```Java
        assertRequest("select * from trades", "18\r\n{\"dataset\":[],\"count\":0}");
        assertRequest("insert into trades values ('2023-07-01T00:00:00.000000', 'ABC', 1.0)", DML_OK);
        assertRequest("insert into trades values ('2023-07-02T00:00:00.000000', 'DEF', 2.0)", DML_OK);
        assertRequest("update trades set value = value + 1 where ticker = 'ABC'", "18\r\n{\"dml\":\"OK\",\"updated\":" + (isWal ? 3 : 1) + "}");

        assertRequest("rename table trades to old_trades", DDL_OK);
        assertGrant(engine, accessListStore, TEST_USER, createPermissions(
                createDBLevelPermissionModel(HTTP, false),
                createTableLevelPermissionModel("old_trades", ALL_TABLE & ~UPDATE, true),
                createColumnLevelPermissionModel("old_trades", new String[]{"ticker", "value"}, UPDATE)
        ));

        assertRequest("alter table old_trades set param maxUncommittedRows =  100000", DDL_OK);
        assertRequest("alter table old_trades set type wal", DDL_OK);
        assertRequest("alter table old_trades set type bypass wal", DDL_OK);
        assertRequest("alter table old_trades set type wal", DDL_OK);
        if (isWal) {
            //set type is applied on restart so suspend/resume is not yet allowed for table that starts as non-wal
            assertRequest("alter table old_trades suspend wal", DDL_OK);
            assertRequest("alter table old_trades resume wal", DDL_OK);
        }
        assertRequest("alter table old_trades detach partition list '2023-07-01'", DDL_OK);

        Files.rename(
                Path.getThreadLocal(root).concat("db").concat("old_trades").concat("2023-07-01.detached").$(),
                Path.getThreadLocal2(root).concat("db").concat("old_trades").concat("2023-07-01.attachable").$()
        );

        assertRequest("alter table old_trades attach partition list '2023-07-01'", DDL_OK);
        assertRequest("alter table old_trades drop partition list '2023-07-01'", DDL_OK);
        assertRequest("backup table old_trades", DDL_OK);
        assertRequest("grant select on old_trades to jane", DDL_OK);
```

Error:

```
2024-11-27T11:24:42.560688Z E i.q.g.e.QueryProgress err [id=-1, sql=`backup table old_trades`, principal=testUser, cache=false, jit=true, time=7308792, msg=cached query plan cannot be used because table schema has changed [table='old_trades', expectedTableId=10, actualTableId=10, expectedMetadataVersion=0, actualMetadataVersion=1], errno=0, pos=0]
2024-11-27T11:24:42.560764Z C i.q.c.h.p.JsonQueryProcessorState [1546188408916] internal error [ex=
io.questdb.cairo.sql.TableReferenceOutOfDateException: cached query plan cannot be used because table schema has changed [table='old_trades', expectedTableId=10, actualTableId=10, expectedMetadataVersion=0, actualMetadataVersion=1]
	at io.questdb.cairo.sql.TableReferenceOutOfDateException.of(TableReferenceOutOfDateException.java:64)
	at io.questdb.cairo.CairoEngine.getReader(CairoEngine.java:562)
	at io.questdb.griffin.SqlExecutionContext.getReader(SqlExecutionContext.java:111)
	at io.questdb.cairo.AbstractPartitionFrameCursorFactory.getReader(AbstractPartitionFrameCursorFactory.java:78)
	at io.questdb.cairo.FullFwdPartitionFrameCursorFactory.getCursor(FullFwdPartitionFrameCursorFactory.java:42)
	at io.questdb.griffin.engine.table.AbstractPageFrameRecordCursorFactory.initPageFrameCursor(AbstractPageFrameRecordCursorFactory.java:96)
	at io.questdb.griffin.engine.table.AbstractPageFrameRecordCursorFactory.getCursor(AbstractPageFrameRecordCursorFactory.java:70)
	at io.questdb.griffin.engine.table.PageFrameRecordCursorFactory.getCursor(PageFrameRecordCursorFactory.java:41)
	at io.questdb.griffin.engine.QueryProgress.getCursor(QueryProgress.java:195)
	at com.questdb.cairo.SecurityCheckFactory.getCursor(SecurityCheckFactory.java:69)
	at io.questdb.griffin.SqlCompilerImpl$DatabaseBackupAgent.backupTable(SqlCompilerImpl.java:3544)
	at io.questdb.griffin.SqlCompilerImpl$DatabaseBackupAgent.sqlTableBackup(SqlCompilerImpl.java:3703)
	at io.questdb.griffin.SqlCompilerImpl$DatabaseBackupAgent.sqlBackup(SqlCompilerImpl.java:3622)
	at io.questdb.griffin.SqlCompilerImpl.lambda$registerKeywordBasedExecutors$0(SqlCompilerImpl.java:3318)
	at io.questdb.griffin.SqlCompilerImpl.compileInner(SqlCompilerImpl.java:1792)
	at io.questdb.griffin.SqlCompilerImpl.compile(SqlCompilerImpl.java:307)
	at io.questdb.cairo.pool.SqlCompilerPool$C.compile(SqlCompilerPool.java:136)
	at io.questdb.cutlass.http.processors.JsonQueryProcessor.compileAndExecuteQuery(JsonQueryProcessor.java:532)
	at io.questdb.cutlass.http.processors.JsonQueryProcessor.execute0(JsonQueryProcessor.java:223)
	at io.questdb.cutlass.http.processors.JsonQueryProcessor.onRequestComplete(JsonQueryProcessor.java:295)
	at io.questdb.cutlass.http.HttpConnectionContext.handleClientRecv(HttpConnectionContext.java:928)
	at io.questdb.cutlass.http.HttpConnectionContext.handleClientOperation(HttpConnectionContext.java:269)
	at io.questdb.cutlass.http.HttpServer.handleClientOperation(HttpServer.java:321)
	at io.questdb.cutlass.http.HttpServer$1.lambda$$0(HttpServer.java:115)
	at io.questdb.network.AbstractIODispatcher.processIOQueue(AbstractIODispatcher.java:199)
	at io.questdb.cutlass.http.HttpServer$1.run(HttpServer.java:119)
	at io.questdb.mp.Worker.run(Worker.java:152)
, q=`backup table old_trades`]
```

### QuestDB version:

8.2

### OS, in case of Docker specify Docker and the Host OS:

MacOS

### File System, in case of Docker specify Host File System:

na

### Full Name:

Alex Pelagenko

### Affiliation:

QuestDB

### Have you followed Linux, MacOs kernel configuration steps to increase Maximum open files and Maximum virtual memory areas limit?

- [X] Yes, I have

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ideoma**
fixes #5201

### CODE REVIEW

**Comment by questdb-butler** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Please update the PR title to match this format:
`type(subType): description`

Where `type` is one of:
`feat`, `fix`, `chore`, `docs`, `style`, `refactor`, `perf`, `test`, `ci`, `revert`

And: `subType` is one of:
`build`, `sql`, `log`, `mig`, `core`, `ilp`, `pgwire`, `http`, `conf`, `ui`, `wal`

For Example:

```
perf(sql): improve pattern matching performance for SELECT sub-queries
```
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against eb0e1113d096a1a8ff7cd9f2c205de4b8f55f02b
</p>

File: core/src/test/resources/sqllogictest/test/sql/test_alter_backup_wal_table.test [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# description: Test dedup cases with string column type
```
Comment (puzpuzpuz) [inline core/src/test/resources/sqllogictest/test/sql/test_alter_backup_wal_table.test:1]: nit: the description is probably copy-paste

**Review by puzpuzpuz — APPROVED** [review]

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 7 / 10 (70.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|7|10|70.00%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 71 / 89 (79.78%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/WaitWalTableFunctionFactory.java|2|3|66.67%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|7|10|70.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|57|70|81.43%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|

**Review by puzpuzpuz — APPROVED** [review]

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 69 / 88 (78.41%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/SqlOptimiser.java|1|2|50.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCompilerImpl.java|7|10|70.00%|
|:large_blue_circle:|io/questdb/cairo/wal/ApplyWal2TableJob.java|57|72|79.17%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|4|4|100.00%|

**Review by puzpuzpuz — APPROVED** [review]

**Pull request merged**